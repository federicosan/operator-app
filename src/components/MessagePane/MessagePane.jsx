import React, { Component, PropTypes } from 'react';
import { connect } from 'react-redux';

// I dislike this imports name so hard :E
import 'whatwg-fetch';

import { NonIdealState } from '@blueprintjs/core';


// TODO: Move this to somewhere that makes sense
const API_URI = 'http://localhost:8000';

export class MessagePaneComponent extends Component {
  static propTypes = {
    dispatch: PropTypes.func,
  };

  constructor (props) {
    super(props);

    this.state = {
      currentChat: null,
      messages: [],
    }
  }

  componentDidUpdate () {
    const { activeChat } = this.props;
    const { currentChat } = this.state;

    console.log('UPDATE');

    // Dont make a request to the API if we dont have an active chat,
    //  or the current chat is the same (i.e. we did a update)
    if (activeChat === null || activeChat === currentChat) {
      return;
    }


    const route = `${API_URI}/api/chat/${activeChat}/messages`;

    fetch(route)
      .then(this.apiWillMount)
      .then(this.apiParse)
      .then(this.apiDidMount.bind(this))
      .catch(this.apiDidNotMount);
  }

  apiWillMount (response) {
    if (response.status >= 200 && response.status < 300) {
      return response;
    }

    let error = new Error(response.statusText);
    error.response = response;

    throw error;
  }

  // TODO: I feel like this can be simplified and removed
  apiParse (response) {
    return response.json();
  }

  apiDidMount (response) {
    const messages = response.messages || [];

    console.log('XHR SUCCESS', response);

    this.setState({
      currentChat: this.props.activeChat,
      messages,
    })
  }

  apiDidNotMount (response) {
    console.log('FAIL', response);
  }

  render () {
    const { messages } = this.state;
    let desc = <span>Womps! But we have so much to say!</span>;
    let layout = (
      <NonIdealState
        title="No Conversations"
        description={desc}
        visual="chat"
      />
    );


    if (!messages.length) {
      // layout = (
      //   <MessageTopBar />
      //   <MessageList />
      //   <InputBar />
      // );
      layout = null
    }

    return (
      <div id="messages">
        {layout}
      </div>
    );
  }
}

const mapStateToProps = state => ({
  // activeChat: state.chat.active
});

const mapDispatchToProps = dispatch => ({
  dispatch,
});

const MessagePane = connect(
  mapStateToProps,
  mapDispatchToProps,
)(MessagePaneComponent);

export default MessagePane;
