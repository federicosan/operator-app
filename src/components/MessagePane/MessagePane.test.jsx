import React from 'react';

import { shallow } from 'enzyme';

import MessagePane from './MessagePane.jsx';

const store = {
  subscribe: jest.fn(),
  dispatch: jest.fn(),
  getState: jest.fn(() => ({ })),
};

describe('Chat', () => {
  it('matches snapshot', () => {
    const component = shallow(<MessagePane store={store} />);

    expect(component).toMatchSnapshot();
  });
});
