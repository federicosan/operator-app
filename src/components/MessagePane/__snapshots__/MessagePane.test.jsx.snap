exports[`Chat matches snapshot 1`] = `
<MessagePaneComponent
  dispatch={[Function]}
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "subscribe": [Function],
    }
  } />
`;
